File: 2023/day01/main.cpp
// go:build ignore
//
#include <algorithm>
#include <iostream>
#include <iterator>
#include <string>
#include <vector>

using namespace std;

const vector<string> digitStrs = {"one", "two",   "three", "four", "five",
                                  "six", "seven", "eight", "nine"};

short charToDigit(const char c) { return c - '0'; }

short getFirstDigit(const string &s) {
  for (auto it = s.begin(); it != s.end(); it++) {
    if (isdigit(*it)) {
      return charToDigit(*it);
    }
  }

  return 0;
}

short getLastDigit(const string &s) {
  for (auto it = s.rbegin(); it != s.rend(); it++) {
    if (isdigit(*it)) {
      return charToDigit(*it);
    }
  }

  return 0;
}

short getFirstDigitWithParse(const string &s) {
  for (auto it = s.begin(); it != s.end(); it++) {
    if (isdigit(*it)) {
      return charToDigit(*it);
    }

    for (int di = 0; const auto &ds : digitStrs) {
      if (it + ds.length() <= s.end() && ds == string(it, it + ds.length())) {
        return di + 1;
      }
      di++;
    }
  }

  return 0;
}

short getLastDigitWithParse(const string &s) {
  for (auto it = s.rbegin(); it != s.rend(); it++) {
    if (isdigit(*it)) {
      return charToDigit(*it);
    }

    for (int di = 0; const auto &ds : digitStrs) {
      if (it + ds.length() <= s.rend() && equal(ds.rbegin(), ds.rend(), it)) {
        return di + 1;
      }
      di++;
    }
  }

  return 0;
}

int main() {

  string line;
  line.reserve(100);
  long sumPart1 = 0;
  long sumPart2 = 0;

  while (getline(cin, line)) {

    sumPart1 += getFirstDigit(line) * 10 + getLastDigit(line);
    sumPart2 += getFirstDigitWithParse(line) * 10 + getLastDigitWithParse(line);
  }
  cout << "Sum part 1: " << sumPart1 << endl;
  cout << "Sum part 2: " << sumPart2 << endl;

  return 0;
}

File: 2023/day01/main.go
package main

import (
	"fmt"
	"strings"
)

func getFirstDigit(str string) int {
	for i := range str {
		if str[i] >= '0' && str[i] <= '9' {
			return int(str[i] - '0')
		}
	}
	return 0
}

func getLastDigit(str string) int {
	for i := range str {
		if str[len(str)-i-1] >= '0' && str[len(str)-i-1] <= '9' {
			return int(str[len(str)-i-1] - '0')
		}
	}
	return 0
}

func getFirstLastNumberFromLine(str string) (int, int) {
	return getFirstDigit(str), getLastDigit(str)
}

var numberStrs = []string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

func getFirstDigitParsed(str string) int {
	for i := range str {
		if str[i] >= '0' && str[i] <= '9' {
			return int(str[i] - '0')
		}

		for di, ds := range numberStrs {
			if strings.HasPrefix(str[i:], ds) {
				return di + 1
			}
		}
	}
	return 0
}

func getLastDigitParsed(str string) int {
	for i := range str {
		if str[len(str)-i-1] >= '0' && str[len(str)-i-1] <= '9' {
			return int(str[len(str)-i-1] - '0')
		}

		for di, ds := range numberStrs {
			if len(str)-i-len(ds) >= 0 && strings.HasPrefix(str[len(str)-i-len(ds):], ds) {
				return di + 1
			}
		}
	}
	return 0
}

func getFirstLastNumberFromLineParsed(str string) (int, int) {
	return getFirstDigitParsed(str), getLastDigitParsed(str)
}

func main() {
	var sum1, sum2 int

	for {
		var s string
		_, err := fmt.Scanln(&s)
		if err != nil {
			break
		}

		first, last := getFirstLastNumberFromLine(s)
		sum1 += first*10 + last

		first2, last2 := getFirstLastNumberFromLineParsed(s)
		sum2 += first2*10 + last2
	}

	fmt.Println("Sum1: ", sum1)
	fmt.Println("Sum2: ", sum2)
}
